File: objectModel/Python/cdm/persistence/cdmfolder/entity_attribute_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from cdm.objectmodel import CdmCorpusContext, CdmEntityAttributeDefinition
from cdm.objectmodel.projections.cardinality_settings import CardinalitySettings
from cdm.enums import CdmObjectType
from cdm.utilities import ResolveOptions, CopyOptions, copy_data_utils, logger
from cdm.enums import CdmLogCode

from . import utils
from .attribute_resolution_guidance_persistence import AttributeResolutionGuidancePersistence
from .entity_reference_persistence import EntityReferencePersistence
from .projections.projection_persistence import ProjectionPersistence
from .purpose_reference_persistence import PurposeReferencePersistence
from .types import EntityAttribute

_TAG = 'EntityAttributePersistence'


class EntityAttributePersistence:
    @staticmethod
    def from_data(ctx: CdmCorpusContext, data: EntityAttribute) -> CdmEntityAttributeDefinition:
        entity_attribute = ctx.corpus.make_object(CdmObjectType.ENTITY_ATTRIBUTE_DEF, data.name)

        entity_attribute.description = data.description
        entity_attribute.display_name = data.displayName
        entity_attribute.explanation = data.explanation

        if data.get('cardinality'):
            min_cardinality = None
            if data.get('cardinality').get('minimum'):
                min_cardinality = data.get('cardinality').get('minimum')

            max_cardinality = None
            if data.get('cardinality').get('maximum'):
                max_cardinality = data.get('cardinality').get('maximum')

            if not min_cardinality or not max_cardinality:
                logger.error(ctx, _TAG, EntityAttributePersistence.from_data.__name__, None, CdmLogCode.ERR_PERSIST_CARDINALITY_PROP_MISSING)

            if not CardinalitySettings._is_minimum_valid(min_cardinality):
                logger.error(ctx, _TAG, EntityAttributePersistence.from_data.__name__, None,
                             CdmLogCode.ERR_PERSIST_INVALID_MIN_CARDINALITY, min_cardinality)

            if not CardinalitySettings._is_maximum_valid(max_cardinality):
                logger.error(ctx, _TAG, EntityAttributePersistence.from_data.__name__, None,
                             CdmLogCode.ERR_PERSIST_INVALID_MAX_CARDINALITY, max_cardinality)

            if min_cardinality and max_cardinality and CardinalitySettings._is_minimum_valid(min_cardinality) and CardinalitySettings._is_maximum_valid(max_cardinality):
                entity_attribute.cardinality = CardinalitySettings(entity_attribute)
                entity_attribute.cardinality.minimum = min_cardinality
                entity_attribute.cardinality.maximum = max_cardinality

        entity_attribute.is_polymorphic_source = data.get('isPolymorphicSource')

        is_projection = data.get('entity') and not isinstance(data.get('entity'), str) and data.get('entity').get('source')

        if is_projection:
            projection = ProjectionPersistence.from_data(ctx, data.get('entity'))
            projection.owner = entity_attribute

            inline_entity_ref = ctx.corpus.make_object(CdmObjectType.ENTITY_REF, None)
            inline_entity_ref.explicit_reference = projection
            entity_attribute.entity = inline_entity_ref
        else:
            entity_attribute.entity = EntityReferencePersistence.from_data(ctx, data.get('entity'))

        entity_attribute.purpose = PurposeReferencePersistence.from_data(ctx, data.get('purpose'))

        applied_traits = utils.create_trait_reference_array(ctx, data.get('appliedTraits'))
        entity_attribute.applied_traits.extend(applied_traits)

        # Ignore resolution guidance if the entity is a projection
        if data.get('resolutionGuidance') and is_projection:
            logger.error(ctx, _TAG, 'from_data', None, CdmLogCode.ERR_PERSIST_ENTITY_ATTR_UNSUPPORTED, entity_attribute.name)
        else:
            entity_attribute.resolution_guidance = AttributeResolutionGuidancePersistence.from_data(ctx, data.get('resolutionGuidance'))

        return entity_attribute

    @staticmethod
    def to_data(instance: CdmEntityAttributeDefinition, res_opt: ResolveOptions, options: CopyOptions) -> EntityAttribute:
        entity = EntityReferencePersistence.to_data(instance.entity, res_opt, options) if instance.entity else None

        entity_attribute = EntityAttribute()
        entity_attribute.name = instance.name
        entity_attribute.description = instance.description
        entity_attribute.displayName = instance.display_name
        entity_attribute.explanation = instance.explanation
        entity_attribute.isPolymorphicSource = instance.is_polymorphic_source
        entity_attribute.purpose = PurposeReferencePersistence.to_data(instance.purpose, res_opt, options) if instance.purpose else None
        entity_attribute.entity = entity
        entity_attribute.appliedTraits = copy_data_utils._array_copy_data(res_opt, instance.applied_traits, options)
        entity_attribute.resolutionGuidance = AttributeResolutionGuidancePersistence.to_data(
            instance.resolution_guidance, res_opt, options) if instance.resolution_guidance else None

        return entity_attribute

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/EntityAttributePersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder;

import java.util.stream.Collectors;

import com.fasterxml.jackson.databind.JsonNode;
import com.microsoft.commondatamodel.objectmodel.cdm.*;
import com.microsoft.commondatamodel.objectmodel.cdm.projections.CardinalitySettings;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmPropertyName;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.projections.ProjectionPersistence;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.EntityAttribute;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.StringUtils;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.Logger;

public class EntityAttributePersistence {

  private static final String TAG = EntityAttributePersistence.class.getSimpleName();

  public static CdmEntityAttributeDefinition fromData(final CdmCorpusContext ctx, final JsonNode obj) {
    final CdmEntityAttributeDefinition entityAttribute =
        ctx.getCorpus().makeObject(CdmObjectType.EntityAttributeDef,
            obj.get("name").asText());


    entityAttribute.setExplanation(Utils.propertyFromDataToString(obj.get("explanation")));
    entityAttribute.updateDescription(Utils.propertyFromDataToString(obj.get("description")));
    entityAttribute.updateDisplayName(Utils.propertyFromDataToString(obj.get("displayName")));

    if (obj.get("cardinality") != null) {
      String minCardinality = null;
      if (obj.get("cardinality").get("minimum") != null) {
        minCardinality = obj.get("cardinality").get("minimum").asText();
      }

      String maxCardinality = null;
      if (obj.get("cardinality").get("maximum") != null) {
        maxCardinality = obj.get("cardinality").get("maximum").asText();
      }

      if (StringUtils.isNullOrTrimEmpty(minCardinality) || StringUtils.isNullOrTrimEmpty(maxCardinality)) {
        Logger.error(ctx, TAG, "fromData", null, CdmLogCode.ErrPersistCardinalityPropMissing);
      }

      if (!CardinalitySettings.isMinimumValid(minCardinality)) {
        Logger.error(ctx, TAG, "fromData", null, CdmLogCode.ErrPersistInvalidMinCardinality, minCardinality);
      }

      if (!CardinalitySettings.isMaximumValid(maxCardinality)) {
        Logger.error(ctx, TAG, "fromData", null, CdmLogCode.ErrPersistInvalidMaxCardinality, maxCardinality);
      }

      if (!StringUtils.isNullOrTrimEmpty(minCardinality) &&
          !StringUtils.isNullOrTrimEmpty(maxCardinality) &&
          CardinalitySettings.isMinimumValid(minCardinality) &&
          CardinalitySettings.isMaximumValid(maxCardinality)) {
        entityAttribute.setCardinality(new CardinalitySettings(entityAttribute));
        entityAttribute.getCardinality().setMinimum(minCardinality);
        entityAttribute.getCardinality().setMaximum(maxCardinality);
      }
    }

    if (obj.has("isPolymorphicSource")) {
      entityAttribute.setIsPolymorphicSource(obj.get("isPolymorphicSource").asBoolean());
    }

    boolean isProjection = obj.get("entity") != null && !(obj.get("entity").isValueNode()) && obj.get("entity").get("source") != null;

    if (isProjection) {
      CdmEntityReference inlineEntityRef = ctx.getCorpus().makeObject(CdmObjectType.EntityRef, null);
      inlineEntityRef.setExplicitReference(ProjectionPersistence.fromData(ctx, obj.get("entity")));
      entityAttribute.setEntity(inlineEntityRef);
    } else {
      entityAttribute.setEntity(EntityReferencePersistence.fromData(ctx, obj.get("entity")));
    }

    entityAttribute.setPurpose(PurposeReferencePersistence.fromData(ctx, obj.get("purpose")));
    Utils.addListToCdmCollection(entityAttribute.getAppliedTraits(),
        Utils.createTraitReferenceList(ctx, obj.get("appliedTraits")));
    // Ignore resolution guidance if the entity is a projection
    if (obj.get("resolutionGuidance") != null && isProjection) {
      Logger.error(ctx, TAG, "fromData", null, CdmLogCode.ErrPersistEntityAttrUnsupported,  entityAttribute.getName());
    } else {
      entityAttribute.setResolutionGuidance(AttributeResolutionGuidancePersistence.fromData(ctx, obj.get("resolutionGuidance")));
    }
    return entityAttribute;
  }

  public static EntityAttribute toData(final CdmEntityAttributeDefinition instance, final ResolveOptions resOpt,
                                       final CopyOptions options) {
    final EntityAttribute result = new EntityAttribute();

    result.setExplanation(instance.getExplanation());
    result.setDescription((String)instance.fetchProperty(CdmPropertyName.DESCRIPTION));
    result.setDisplayName((String)instance.fetchProperty(CdmPropertyName.DISPLAY_NAME));
    result.setName(instance.getName());
    result.setIsPolymorphicSource(instance.getIsPolymorphicSource());
    result.setEntity(Utils.jsonForm(instance.getEntity(), resOpt, options));
    result.setPurpose(Utils.jsonForm(instance.getPurpose(), resOpt, options));
    result.setAppliedTraits(Utils.listCopyDataAsArrayNode(
      instance.getAppliedTraits().getAllItems()
          .stream()
          .filter(trait -> trait instanceof CdmTraitGroupReference || !((CdmTraitReference)trait).isFromProperty())
          .collect(Collectors.toList()),
      resOpt,
      options));
    result.setResolutionGuidance(Utils.jsonForm(instance.getResolutionGuidance(), resOpt, options));
    return result;
  }
}
